Add Apple OAuth2 authentication support

silva-fj opened the pull request:
## Summary

Add Apple OAuth2 authentication support and refactor OAuth2 flow.

### Changes

- Add Apple as OAuth2 provider (alongside existing Google support)
- Add Apple identity type to primitives (Identity, UserId, Web2IdentityType)

### OAuth2 Flow Refactoring (Implicit Flow)

**Security enhancements:**
- Add nonce generation and verification (prevents replay attacks)
- Verify id_token from client before code exchange (aud + nonce checks)
- Store both state and nonce in `OAuth2VerificationData`

**API changes:**
- Rename: `omni_getOAuth2AuthorizationUrl` → `omni_getOAuth2AuthorizationData`
- Return structured data object (url, state, nonce, scope, etc.) instead of just URL
- Add `id_token` parameter to `OAuth2Data`
- Remove Identity/sender from `OmniAuth::OAuth2` variant


### Environment Variables
```bash
OE_APPLE_CLIENT_ID_{CLIENT}=<client_id>
OE_APPLE_CLIENT_SECRET_{CLIENT}=<client_secret>
```

silva-fj committed:
Add Apple identity type to primitives

- Add Apple variant to Identity enum (codec index 11)
- Add Apple to Web2IdentityType enum
- Add Apple to UserId enum with conversions
- Add Apple variant to OAuth2Provider enum
- Update identity helpers (to_native_account, to_omni_account, to_did)

silva-fj committed:
Add Apple OAuth2 configuration infrastructure

- Add AppleOAuth2Config struct to config-loader
- Implement env var loading (OE_APPLE_CLIENT_ID_*, OE_APPLE_CLIENT_SECRET_*)
- Add AppleOAuth2Factory with caching support
- Export AppleOAuth2Config in public API

silva-fj committed:
Add Apple OAuth2 client and helpers

- Implement AppleOAuth2Client for token exchange
- Add Apple authorization URL helper with scope 'email'
- Add ID token decoder for Apple tokens
- Export apple module from oauth-providers and identity-verification

silva-fj committed:
Refactor OAuth2 RPC methods to support Apple

- Rename omni_getOAuth2GoogleAuthorizationUrl -> omni_getOAuth2AuthorizationUrl
- Add provider parameter to support both Google and Apple
- Update login_with_oauth2 to accept Apple provider
- Add Apple factory to RpcContext
- Update error messages to include both providers

silva-fj committed:
Add Apple OAuth2 verification logic

- Implement verify_apple_oauth2 function
- Add state verifier validation for Apple
- Add code-to-token exchange for Apple
- Extract email from Apple ID token
- Update verify_oauth2_authentication to handle Apple provider

silva-fj committed:
Add Copy trait to OAuth2Provider enum

Makes OAuth2Provider more ergonomic to use by avoiding unnecessary clones

silva-fj committed:
Refactor OAuth2 configuration to eliminate duplication

- Merge GoogleOAuth2Config and AppleOAuth2Config into unified OAuth2Config
- Replace separate config maps with single oauth2_configs map
- Create generic load_oauth2_configs function for all providers
- Add unified get_oauth2_config(client_id, provider) method
- Reduce config loading code by ~100 lines

silva-fj committed:
Create unified OAuth2Client and provider config trait

- Add OAuth2ProviderConfig trait for provider-specific settings
- Implement GoogleProviderConfig and AppleProviderConfig
- Create unified OAuth2Client replacing duplicate implementations
- Remove separate google.rs and apple.rs files (110 lines eliminated)

silva-fj committed:
Create shared OAuth2 authorization helpers

- Add oauth2_common module with shared AuthorizeData struct
- Create generic get_authorize_data function
- Refactor Google and Apple modules to use shared logic
- Eliminate ~40 lines of duplicated authorization code

silva-fj committed:
Replace separate factories with unified OAuth2ConfigFactory

- Create OAuth2ConfigFactory supporting all providers
- Remove GoogleOAuth2Factory and AppleOAuth2Factory (157 lines deleted)
- Unified caching with (client_id, provider) key
- Single get_config method for all providers

silva-fj committed:
Unify OAuth2 verification logic into single function

- Replace verify_google_oauth2 and verify_apple_oauth2 with verify_oauth2_provider
- Eliminate 80+ lines of duplicated verification code
- Share state validation, config fetching, and token exchange logic
- Provider-specific handling only for token decoding

silva-fj committed:
Simplify RPC layer with unified OAuth2 factory

- Update get_oauth2_authorization_url to use single factory
- Replace two factory fields in RpcContext with one oauth2_factory
- Cleaner initialization with single factory instantiation
- Streamline authorization URL generation logic

silva-fj committed:
feat(identity): add Apple variant to token conversion

silva-fj committed:
Add Apple identity variant support to identity worker

- Add Apple case to prometheus metrics identity match
- Add Apple variant to OAuth2Provider enum in authentication_utils
- Add placeholder verification (returns not yet supported error)

silva-fj committed:
Add Apple variant to parachain identity tests

- Fix 6 exhaustive match patterns in test functions
- Apple is web2 (true), not web3/substrate/evm/bitcoin/solana (false)

BillyWooo commented on tee-worker/omni-executor/heima/identity-verification/src/web2/apple/mod.rs:
`apple/mod.rs` and `google/mod.rs` are almost the same (Only the different `IdToken` definition, right?). Can we merge them to one trait implementation?

BillyWooo commented on tee-worker/omni-executor/rpc-server/src/methods/omni/login_with_oauth2.rs:
the `jwt` token created here is of type `AUTH_TOKEN_ACCESS_TYPE`. But in `verify_auth()`, it's verifying against `AUTH_TOKEN_ID_TYPE`.  Because later the jwt will be used to do `verify_auth()` when attaching passkey. Any solution?

silva-fj replied:
Ah yeah, I forgot we issue two types of token. It's fixed now 👍🏼

silva-fj committed:
fix(auth): use correct token type for OAuth2 login

silva-fj committed:
Extract shared JWT token decoding to oauth2_common

- Move decode_id_token logic to oauth2_common as generic function
- Move base64_decode helper to oauth2_common
- Update Apple and Google modules to use shared implementation
- Eliminate 28 lines of duplicated JWT decoding code

silva-fj committed:
feat(oauth2): add nonce to OAuth2 verification flow

- Add OAuth2VerificationData struct to store state and nonce
- Add id_token field to OAuth2Data for client-side token verification
- Remove Identity parameter from OmniAuth::OAuth2 variant
- Update OAuth2StateVerifierStorage to store OAuth2VerificationData

silva-fj committed:
fix(oauth2): correct IdToken field types for Google and Apple

Google IdToken:
- email_verified: bool -> String (Google returns "true"/"false")
- hd: String -> Option<String> (only for Workspace accounts)
- at_hash: String -> Option<String>
- nonce: String -> Option<String>

Apple IdToken:
- email_verified: Option<String> -> bool
- nonce: String -> Option<String>
- Add optional fields: is_private_email, c_hash, auth_time

silva-fj committed:
refactor(oauth2): add OAuth2ProviderConfig helper

- Add OAuth2ProviderConfig struct to centralize provider configuration
- Export BASE_URL and SCOPES as public constants in provider modules
- Remove wrapper functions (get_authorize_data, decode_id_token)
- Add nonce parameter to authorization data

silva-fj committed:
refactor(rpc): rename OAuth2 authorization method and return data object

- Rename omni_getOAuth2AuthorizationUrl -> omni_getOAuth2AuthorizationData
- Return structured data (url, state, nonce, scope, response_type, etc)
- Use OAuth2ProviderConfig for cleaner provider-specific configuration
- Store both state and nonce in OAuth2VerificationData

silva-fj committed:
feat(oauth2): verify id_token with nonce and audience checks

- Add id_token parameter to login flow
- Verify id_token before token exchange (decode, aud, nonce)
- Add verify_id_token_claims helper to reduce code duplication
- Remove Identity parameter from OAuth2 verification
- Use oauth2_common::decode_id_token for both providers

silva-fj committed:
fix(oauth2): change Google IdToken email_verified to bool type

BillyWooo reviewed (approved):
@Kailai-Wang  @kziemianek Please also take a look. Thank you.

silva-fj committed:
refactor(oauth2): make redirect_uri optional in OAuth2Data

silva-fj committed:
refactor(oauth2): make redirect_uri optional in helper functions

silva-fj committed:
refactor(oauth2): make redirect_uri optional in RPC methods

BillyWooo committed:
Merge branch 'dev' into p-1707-login-with-apple

silva-fj committed:
test(oauth2): fix tests for optional redirect_uri
